### PR TITLE
Block Deprecations: Provide extra data for isEligible check

### PR DESCRIPTION
**Pull request by aaronrobertshaw**
Related:
- https://github.com/WordPress/gutenberg/issues/20941
- https://github.com/WordPress/gutenberg/pull/48404

## What?

Provides additional data from which deprecations can check if they are eligible.

## Why?

The information currently passed to the `isEligible` check is insufficient in some cases to determine if a deprecation should in fact run. 

One use case requiring additional information can be found in:
- https://github.com/WordPress/gutenberg/pull/48404

In the above example, when the default `align` attribute has its default switched to `none`, existing blocks using the default `wide` align are not determined to be invalid. This is due to the custom CSS class name support picking up the deprecated `alignwide` class from the markup and adding it to the `className` attribute.

As the deprecated block is still determined to be valid, the deprecation's `isEligible` callback must be able to return true in this scenario. The catch is presently the `isEligible` function is passed only the raw parsed attributes and inner blocks. The raw parsed attributes do not yet have any amendments made to the `className` attribute by the custom class names support.

## How?

- Passes a third param to a deprecation's `isEligible` function. That being and object now including the current block e.g. `{ block }`.
- An object was chosen as it could more easily be extended if there is a need for access to the raw markup, raw block etc.
- No changes are required to existing `isEligible` function definitions.

#### Pass only `rawBlock` and current `block` as extra params

This would avoid the creation of a new object and potential minimise any performance hit. It would make for a rather awkward signature and API.

#### `isEligibleV2`

This approach would allow deprecations to define their eligibility check via a new function signature.
e.g. `isEligibleV2( rawBlock, currentBlock )`.

This would still provide the existing data via `rawBlock.attributes` and `currentBlock.innerBlocks` but also allow checks to be made against the markup or finalized block attributes. 

Some downsides are the multiple function signatures, the potential need to deprecate one eventually, potential confusion etc. The upside is primarily the clean function signature and only passing the extra data when a deprecation desires it.

#### Adding a disallow list to Custom CSS Classname support

This is a more heavy-handed and hacky approach with a greater potential performance hit.

Blocks could define a disallow list for custom class names. That list can be checked during the determination of custom class names from the block's markup. A CSS class is only added to the `className` attribute if it wasn't in the disallow list.

Using the media and text example, `alignwide` could be disallowed, which would then cause unmigrated blocks to be invalidated triggering the necessary deprecation without the need for the `isEligible` check.

## Testing Instructions
1. Ensure block fixture tests still pass: `npm run test:unit test/integration/full-content/full-content.test.js`
2. Proofread block deprecation docs update
3. Test https://github.com/WordPress/gutenberg/pull/48404 which is now based on this PR

### CODE REVIEW

**Comment by aaronrobertshaw** [comment]
#48404 has been rebased to leverage this update via https://github.com/WordPress/gutenberg/pull/48404/commits/82da6db2393010295fe390719af5e82b42ab0308.

**Comment by aaronrobertshaw** [comment]
@youknowriad would you mind weighing in on this change or suggesting a better alternative?

**Comment by youknowriad** [comment]
I think we've found in the past that changing `default` values has a lot of implications and not just for deprecations and should probably be avoid. My memory is failing me though, maybe it was server rendering related, but I'm not entirely sure.

**Review by youknowriad — COMMENTED** [review]
To be honest, I don't like the current `isEligible` API, and I'd say that our deprecation behavior/algorithm is not perfect at all but this is going to be a huge effort to improve. Now passing the block like done in this PR feels a bit random. Maybe a new API is a bit better if this proves to be really necessary.

Maybe others have thoughts here too @gziolo @dmsnell

**Comment by aaronrobertshaw** [comment]
Thanks @youknowriad 👍 

> My memory is failing me though, maybe it was server rendering related, but I'm not entirely sure.

Could this [PR](https://github.com/WordPress/gutenberg/pull/42763) around the layout `default` for the Group block be one of the past issues you're referring to?

At first glance, changing the default layout for the Group block definitely had wider ranging impacts. The media and text block default alignment seemed reduced in scope but I'm happy to be warned of unconsidered pitfalls though.

> To be honest, I don't like the current isEligible API, and I'd say that our deprecation behavior/algorithm is not perfect at all but this is going to be a huge effort to improve. Now passing the block like done in this PR feels a bit random.

100% agreed.

**Review by dmsnell — COMMENTED** [review]
Thanks for the ping @youknowriad. Sharing my approval here optimistically, to note my approval of the idea, not to signal that you ought to merge this without the review and feedback from others 😉 

> Now passing the block like done in this PR feels a bit random. Maybe a new API is a bit better if this proves to be really necessary.

Given that deprecations already have trouble with the designed API I'm more inclined to let the extra parameters in. To that end I think `rawBlock` is more appropriate than `block` because of the implications of `block` at this point.

We can call `rawBlock` the `blockNode` (hat-tip to @mcsf who I think coined that term) or the "stage-1 parse of the block," which is what we get when parsing the serialized HTML.

The `block` object here, or the "stage-2 parse," is the result of feeding that block node into the code for the given block type.

The distinction _might_ be important here because we're in a situation where we know _something_ didn't load as expected and that `block` is likely invalid (according to what the editor calls "invalid"). It can be misleading because it may have lost correspondence with the serialized HTML. For example, the list block updated to use inner blocks, so old list blocks will have the `innerHTML` in that block node with the `<ul>` (or `<ol>`) and `<li>` tags, but will render an empty list into `block`, because there are no inner blocks.

A deprecation would need to restore the original content that in a fraction of the cases will be wiped out before running the deprecation code.

> our deprecation behavior/algorithm is not perfect at all but this is going to be a huge effort to improve

this is another reason that being loose here feels favorable: an overhaul of the block updating and validation-handling code might wipe out a number of these fixes through very different means and I suspect there's little likelihood we can anticipate now how those will appear later.

**Comment by aaronrobertshaw** [comment]
Thanks for weighing in @dmsnell you make some great points 👍 

It still appears to me that there could be some benefit in proceeding with the addition of the extra args to `isEligible` and that it might be better to include both the `blockNode` and `block` objects.  I've updated this PR to do that.

I don't think there is a huge rush on this so it can wait for sufficient reviews for everyone to be confident in the change.

File: docs/reference-guides/block-api/block-deprecation.md
Comment (talldan) [inline docs/reference-guides/block-api/block-deprecation.md]: What do you think about re-structuring the text for `isEligible` (and maybe `migrate` too)?

A few changes that I think would be good:
- Document the arguments using bullet points (matching what we do for function that have JSDoc types).
- Detail the expected return type (again like we do for JSDoc types).
- Ensure the terms we're using are consistent (e.g. parsed block and block node are both mentioned currently).
- Add some more clarity to that last sentence. I've read it a few times and still don't understand it.

Ideally `isEligible` and `migrate` are so complicated as concepts that they should be in a separate section, but that's not one for this PR.

Comment (aaronrobertshaw) [reply docs/reference-guides/block-api/block-deprecation.md]: Some great suggestions there, thank you 👍 

I've tried reformatting and rewording the `isEligible` and `migrate` function docs in https://github.com/WordPress/gutenberg/pull/48815/commits/efd2cd7bd6f484ada942433ed4fc32dc4786e261. 

Let me know if any further tweaks are needed in this PR.

File: docs/reference-guides/block-api/block-deprecation.md
Comment (talldan) [inline docs/reference-guides/block-api/block-deprecation.md]:
```suggestion
		- `attributes`: The raw block attributes as parsed from the serialized HTML, and before the block type code is applied.
```

File: docs/reference-guides/block-api/block-deprecation.md
Comment (talldan) [inline docs/reference-guides/block-api/block-deprecation.md]:
```suggestion
			- `data.blockNode`: The raw form of the block as a result of parsing the serialized HTML.
```

File: docs/reference-guides/block-api/block-deprecation.md
Comment (talldan) [inline docs/reference-guides/block-api/block-deprecation.md]:
```suggestion
			- `data.block`: The block object, which is the result of applying the block type to the `blockNode`.
```

**Review by talldan — APPROVED** [review]
Thanks for reworking the docs.

I left a few suggestions for improving consistency, as there were a few different terms being used (like 'serialized HTML' vs 'saved markup') that mean the same thing, but I leave these up to you.

**Comment by aaronrobertshaw** [comment]
> I left a few suggestions for improving consistency, as there were a few different terms being used (like 'serialized HTML' vs 'saved markup') that mean the same thing, but I leave these up to you.

Thanks. The suggestions look good. I've committed those and will merge once the e2es pass.

**Pull request merged**
